src/sql.rs

```diff
@@ -0,0 +1,1175 @@
+    value: i64,
+    field2: impl AsRef<str>,
+    value2: i32,
+) -> u32 {
+    sql.with_conn(|conn| {
+        Ok(get_rowid2_with_conn(
+            context, conn, table, field, value, field2, value2,
+        ))
+    })
+    .unwrap_or_else(|_| 0)
+}
+
+pub fn get_rowid2_with_conn(
+    context: &Context,
+    conn: &Connection,
+    table: impl AsRef<str>,
+    field: impl AsRef<str>,
+    value: i64,
+    field2: impl AsRef<str>,
+    value2: i32,
+) -> u32 {
+    match conn.query_row(
+        &format!(
+            "SELECT id FROM {} WHERE {}={} AND {}={} ORDER BY id DESC",
+            table.as_ref(),
+            field.as_ref(),
+            value,
+            field2.as_ref(),
+            value2,
+        ),
+        NO_PARAMS,
+        |row| row.get::<_, u32>(0),
+    ) {
+        Ok(id) => id,
+        Err(err) => {
+            error!(context, 0, "sql: Failed to retrieve rowid2: {}", err);
+            0
+        }
+    }
+}
+
+pub fn housekeeping(context: &Context) {
+    let mut files_in_use = HashSet::new();
+    let mut unreferenced_count = 0;
+
+    info!(context, 0, "Start housekeeping...");
+    maybe_add_from_param(
+        context,
+        &mut files_in_use,
+        "SELECT param FROM msgs  WHERE chat_id!=3   AND type!=10;",
+        'f' as i32,
+    );
+    maybe_add_from_param(
+        context,
+        &mut files_in_use,
+        "SELECT param FROM jobs;",
+        'f' as i32,
+    );
+    maybe_add_from_param(
+        context,
+        &mut files_in_use,
+        "SELECT param FROM chats;",
+        'i' as i32,
+    );
+    maybe_add_from_param(
+        context,
+        &mut files_in_use,
+        "SELECT param FROM contacts;",
+        'i' as i32,
+    );
+
+    context
+        .sql
+        .query_map(
+            "SELECT value FROM config;",
+            params![],
+            |row| row.get::<_, String>(0),
+            |rows| {
+                for row in rows {
+                    maybe_add_file(&mut files_in_use, row?);
+                }
+                Ok(())
+            },
+        )
+        .unwrap_or_else(|err| {
+            warn!(context, 0, "sql: failed query: {}", err);
+        });
+
+    info!(context, 0, "{} files in use.", files_in_use.len(),);
+    /* go through directory and delete unused files */
+    let p = std::path::Path::new(as_str(context.get_blobdir()));
+    match std::fs::read_dir(p) {
+        Ok(dir_handle) => {
+            /* avoid deletion of files that are just created to build a message object */
+            let diff = std::time::Duration::from_secs(60 * 60);
+            let keep_files_newer_than = std::time::SystemTime::now().checked_sub(diff).unwrap();
+
+            for entry in dir_handle {
+                if entry.is_err() {
+                    break;
+                }
+                let entry = entry.unwrap();
+                let name_f = entry.file_name();
+                let name_c = to_cstring(name_f.to_string_lossy());
+
+                if unsafe {
```

Comment:
You could stuff this whole if statement in an outer unsafe, get rid off all the inner unsafes and I think it'd become more readable.